Sofa/Component/Mapping/NonLinear/src/sofa/component/mapping/nonlinear/AreaMapping.inl

```diff
@@ -0,0 +1,376 @@
+        l_topology.set(this->getContext()->getMeshTopologyLink());
+    }
+
+    msg_info() << "Topology path used: '" << l_topology.getLinkedPath() << "'";
+
+    const auto nbSurfacePrimitives = l_topology->getNbQuads() + l_topology->getNbTriangles();
+
+    if (nbSurfacePrimitives == 0)
+    {
+        msg_error() << "No topology component containing surface primitives found at path: " << l_topology.getLinkedPath() << ", nor in current context: " << this->getContext()->name;
+        this->d_componentState.setValue(sofa::core::objectmodel::ComponentState::Invalid);
+        return;
+    }
+
+    typename core::behavior::MechanicalState<In>::ReadVecCoord pos = this->getFromModel()->readPositions();
+
+    this->getToModel()->resize( nbSurfacePrimitives );
+    jacobian.resizeBlocks(nbSurfacePrimitives, pos.size());
+
+    baseMatrices.resize( 1 );
+    baseMatrices[0] = &jacobian;
+
+    Inherit1::init();
+}
+
+template <class TIn, class TOut>
+void AreaMapping<TIn, TOut>::apply(const core::MechanicalParams* mparams,
+    DataVecCoord_t<Out>& out, const DataVecCoord_t<In>& in)
+{
+    helper::WriteOnlyAccessor< Data<VecCoord_t<Out>> > _out = out;
+    helper::ReadAccessor< Data<VecCoord_t<In> > > _in = in;
+
+    m_vertices = _in.operator->();
+
+    const auto& triangles = l_topology->getTriangles();
+    const auto& quads = l_topology->getQuads();
+
+    jacobian.clear();
+
+    for (unsigned int triangleId = 0; triangleId < triangles.size(); ++triangleId)
+    {
+        const auto& triangle = triangles[triangleId];
+
+        const auto& n0 = TIn::getCPos(_in[triangle[0]]);
+        const auto& n1 = TIn::getCPos(_in[triangle[1]]);
+        const auto& n2 = TIn::getCPos(_in[triangle[2]]);
+
+        const auto n01 = n1 - n0;
+        const auto n02 = n2 - n0;
+        const auto N = sofa::type::cross(n01, n02);
+        const auto norm = N.norm();
+
+        const auto area = static_cast<typename In::Real>(0.5) * norm;
+
+        _out[triangleId] = area;
+
+        const auto k = 1 / (2 * norm);
+
+        sofa::type::fixed_array<JacobianEntry, 3> jacobianEntries {
+            JacobianEntry{triangle[0], k * sofa::type::cross(n1-n2, N)}, // dArea_dn0
+            JacobianEntry{triangle[1], k * sofa::type::cross(n02, N)}, // dArea_dn1
+            JacobianEntry{triangle[2],-k * sofa::type::cross(n01, N)}, // dArea_dn2
+        };
+
+        //insertion in increasing column order
+        std::sort(jacobianEntries.begin(), jacobianEntries.end());
+
+        jacobian.beginRow(triangleId);
+        for (const auto& [vertexId, jacobianValue] : jacobianEntries)
+        {
+            for (unsigned d = 0; d < In::spatial_dimensions; ++d)
+            {
+                jacobian.insertBack(triangleId, vertexId * Nin + d, jacobianValue[d]);
+            }
+        }
+    }
+
+    if (!quads.empty())
+    {
+        msg_warning() << "Quads are not supported in this Mapping. Convert them to triangles first.";
```

Comment:
If the component is not working with quads, it should reports a msg_error & exist with proper componentState

Reply:
done